### PR TITLE
Generic abstractions for virtio devices and the MMIO transport

### PR DESCRIPTION
**Pull request by alexandruag**
This PR proposes generic abstractions for virtio devices and the MMIO transport in particular (the aim is to add PCI support as well at some point). We first redefine the `VirtioDevice` trait, which contains methods required to support interactions that happen at the transport level. The following commit defines `VirtioMmioDevice`, which provides default implementations (based on the functionality from `VirtioDevice`) for reads and writes to the registers/configuration space of a virtio device that uses the MMIO transport. Finally, the last commit introduces the `VirtioConfig`, `WithVirtioConfig`, and `WithDeviceOps` helper abstractions, which are used to provide automatic implementations of the previous traits.

There are more details present in the code comments. This [commit](https://github.com/alexandruag/vm-virtio/commit/32f26359c6fa5996f62ae767115c793f43ab78af#diff-060bce9da11057f730f84bbc0e029eb859f54cc5e05b05bed48f5a8f1d4df524) contains a simple example of how the previous pieces fit together to simplify the implementation of virtio-over-MMIO functionality for a potential device. I'm working on a more complete example of an actual working device and will add it here as well once it's available. Looking forward to any feedback!

### CODE REVIEW

**Comment by alexandruag** [comment]
[Here](https://github.com/alexandruag/vmm-reference/blob/block_rfc/src/devices/src/virtio/block/device.rs)'s a better example of how the previously mentioned abstractions can be used as part of a block device implementation.

File: src/devices/mmio.rs
Comment (aghecenco) [inline src/devices/mmio.rs:114]: +1. It can have a const fn and default implementation, not a big change, and would be neater.

File: src/devices/mmio.rs
Comment (aghecenco) [inline src/devices/mmio.rs:117]: According to the spec: `Device-specific configuration space starts at the offset 0x100 and is accessed with byte alignment.`
Admittedly, `byte alignment` is a weird phrase but looking next door, qemu interprets this as `only byte, word or dword accesses are allowed` ([code](https://github.com/qemu/qemu/blob/c8e5c4b246584da36694a3c259a7dbb8a7e7b1f3/hw/virtio/virtio-mmio.c#L113)). I think this makes sense and we should do the same check, WDYT?

**Comment by slp** [comment]
I like both the design and the implementation. :+1:

File: src/devices/mmio.rs
Comment (rbradford) [reply src/devices/mmio.rs:117]: The byte alignment here means you're not restricted to only accessing it at a word,dword or qword granularity. i.e. you can access it however you like so there is no check you can make.

**Comment by alexandruag** [comment]
Thanks for the comments everyone! I'll address them and turn this into a regular PR over the following days.

File: src/lib.rs
Comment (andreeaflorescu) [inline src/lib.rs]: I think that queues and devices are conceptually different and can be used independently (at least the queues). For this reason, what do you say about exporting the modules directly instead of `pub use`?

My general thinking when using flattened vs non-flattened exports is that flattened exports are to be used when the modules are there just to provide a structural separation, and not a logical one.

**Structural separation**: If everything would be in one module, than the module would be to big, and thus the code becomes hard to follow, hence we are splitting it in multiple modules. This artificial separation is hidden from the end user by exporting everything flattened.
**Logical separation**: separate concepts are split between logical modules.

File: src/devices/mmio.rs
Comment (andreeaflorescu) [inline src/devices/mmio.rs:32]: I deeply dislike these kind of functions that don't belong to any object :(

**Review by andreeaflorescu — COMMENTED** [review]
I just looked briefly over the interfaces. I'll do an in-depth review of the implementation when this is not longer an RFC.

I find the `WithDriverSelection` a bit too strange. I believe we can work on improving the readability of that code by providing a better abstraction (that is in case I understood correctly what you're trying to achieve with it).

Otherwise, the PR LGTM.

File: src/devices/mmio.rs
Comment (alexandruag) [reply src/devices/mmio.rs:114]: On second though, the effective size of the configuration space is actually variable depending on selected features, so I would keep the static upper bound and update TODO for now to think a bit more about this. Ultimately, each specific implementation of `read/write_config` takes care of validations, etc.

File: src/devices/virtio_config.rs
Comment (alexandruag) [reply src/devices/virtio_config.rs:24]: Yeah so like I mentioned in a previous comment, the driver can write a full 32-bit value when selecting a page, but then all the feature bits defined by the standard so far appear to fit within a `u64`. I'll use the simplification mentioned there if it sounds good.

File: src/devices/virtio_config.rs
Comment (alexandruag) [reply src/devices/virtio_config.rs:90]: Yeah so I don't know which name is best here, but at the same time I don't know if `DeviceStateUpdate` fits that well. We can think a bit more and just go with something in the end.

File: src/devices/virtio_config.rs
Comment (alexandruag) [reply src/devices/virtio_config.rs:129]: This method might disappear, depending on the discussion around the previous comment about removing it.

**Comment by alexandruag** [comment]
I've simplified `devices_features`/`driver_features` handling a bit (as per one of the previous comments) and added unit tests in a new, separate commit. I also removed the draft status from the PR. Thanks for all the comments so far and please take another look.

**Comment by alexandruag** [comment]
Did another push that renames some of the helper traits in the fourth commit, uses `BorrowMut` to abstract access to a `VirtioConfig` object, and we're also setting the `event_idx` queue flag as part of feature acknowledgement for extra convenience.

File: src/device/mmio.rs
Comment (aghecenco) [inline src/device/mmio.rs:92]: This can be tested too fairly easily if the `Dummy` device does `d.set_device_features_select(2)` before an MMIO read from offset `0x10`

File: src/device/mmio.rs
Comment (aghecenco) [inline src/device/mmio.rs:163]: This is also rather easily testable if you override `fn queue_notify` for `Dummy` and force a write here.

File: src/device/mod.rs
Comment (aghecenco) [inline src/device/mod.rs:146]: I think there's a way to test this too with `Dummy` though I don't have the specifics.

**Review by aghecenco — APPROVED** [review]
I have some final nits and minor test enhancements that don't look too hard to add, but I think the PR is good to go as-is. Lack of proper documentation is my main note here, but that can wait until the crate is ready for publishing.

**Review by slp — DISMISSED** [review]

**Comment by slp** [comment]
LGTM. I think it can be merged once is rebased.

File: src/device/virtio_config.rs
Comment (alexandruag) [reply src/device/virtio_config.rs:175]: Sure, will add an issue immediately after this gets merged in the current form.

**Review by alexandruag — COMMENTED** [review]
Addressed the latest comments and did a rebase.

**Review by aghecenco — APPROVED** [review]

File: src/device/mod.rs
Comment (jiangliu) [inline src/device/mod.rs:48]: Could we sort these consts by value?

**Review by slp — APPROVED** [review]

File: .buildkite/custom-pipeline.yml
Comment (alexandruag) [reply .buildkite/custom-pipeline.yml:12]: Yeah it makes sense to do so; we can piggy back the update on another PR.

Comment (andreeaflorescu) [reply .buildkite/custom-pipeline.yml:12]: I've opened an issue to update the rust-vmm-ci as well as it is super old: https://github.com/rust-vmm/vm-virtio/issues/34

I'll add a comment to also update the container version in the custom-pipeline.

File: src/device/mod.rs
Comment (alexandruag) [reply src/device/mod.rs:48]: They were initially written down in the order they are presented in the standard (i.e. in section 2.1), but we can def reorder them if it improves clarity & readability.

**Pull request merged**
